core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -231,9 +237,35 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
     }
   }
 
+  private void onSubchannelHealth(HealthUtil.HealthCheckingListener hcListener,
+                                  HealthUtil.HealthStatus hcStatus) {
+    log.log(Level.FINE, "Received health status: " + hcStatus + " listener:" + hcListener);
+    Subchannel find = null;
+    for (SubchannelData subchannelData : subchannels.values()) {
+      if (subchannelData.healthCheckingListener == hcListener
+            && subchannelData.getState() == READY) {
+        find = subchannelData.subchannel;
+        break;
+      }
+    }
+    if (find == null) {
+      return;
+    }
+    if (hcStatus.servingStatus() != HealthUtil.ServingStatus.SERVING) {
+      helper.updateBalancingState(TRANSIENT_FAILURE,
+          new FixedResultPicker(PickResult.withError(
+              Status.UNAVAILABLE.withDescription("unhealthy"))));
+      concludedState = TRANSIENT_FAILURE;
+    } else {
+      helper.updateBalancingState(READY, new Picker(PickResult.withSubchannel(find)));
+      concludedState = READY;
+    }
+  }
+
   private void updateBalancingState(ConnectivityState state, SubchannelPicker picker) {
     if (state != currentState || state == READY || state == TRANSIENT_FAILURE) {
       currentState = state;
+      concludedState = state;
```

Comment:
You have no information here about health states, so no reason to change concludedState.